includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -662,6 +662,12 @@ private function has_used_class_name( $class_names ) {
 				}
 			}
 
+			// If the document has an amp-date-picker tag, check if this class is an allowed child of it.
+			// That component's child classes won't be present yet in the document, so prevent tree-shaking valid classes.
+			if ( $this->has_used_tag_names( [ 'amp-date-picker' ] ) && $this->is_class_allowed_in_amp_date_picker( $class_name ) ) {
```

Comment:
This needs to be integrated into the switch logic for the datepicker above that starts at line 651. Otherwise, we will _always_ run the optimized loop, and then _always_ run another check for the datepicker, suddenly making the entire loop work twice as much (and thus much slower).

Reply:
Thanks, good point that this would be a big performance cost always running this other check.

Maybe I'm not understanding your suggestion. 

But on moving this into the switch for the datepicker at line 651:`

```diff
diff --git a/includes/sanitizers/class-amp-style-sanitizer.php b/includes/sanitizers/class-amp-style-sanitizer.php
index e9e2b4be6..5f807791a 100644
--- a/includes/sanitizers/class-amp-style-sanitizer.php
+++ b/includes/sanitizers/class-amp-style-sanitizer.php
@@ -649,7 +649,7 @@ class AMP_Style_Sanitizer extends AMP_Base_Sanitizer {
                                switch ( substr( $class_name, 0, 16 ) ) {
                                        // Class names for amp-date-picker, see <https://www.ampproject.org/docs/reference/components/amp-date-picker>.
                                        case 'amp-date-picker-':
-                                               if ( ! $this->has_used_tag_names( [ 'amp-date-picker' ] ) ) {
+                                               if ( ! $this->has_used_tag_names( [ 'amp-date-picker' ] ) || ! $this->is_class_allowed_in_amp_date_picker( $class_name ) ) {  // Though this conditional will probably never be false.
                                                        return false;
                                                }
                                                continue 2;
@@ -662,12 +662,6 @@ class AMP_Style_Sanitizer extends AMP_Base_Sanitizer {
                                }
                        }
 
-                       // If the document has an amp-date-picker tag, check if this class is an allowed child of it.
-                       // That component's child classes won't be present yet in the document, so prevent tree-shaking valid classes.
-                       if ( $this->has_used_tag_names( [ 'amp-date-picker' ] ) && $this->is_class_allowed_in_amp_date_picker( $class_name ) ) {
-                               continue;
-                       }
-
                        if ( ! isset( $this->used_class_names[ $class_name ] ) ) {
                                return false;
                        }
```

...this won't allow the classes in `is_class_allowed_in_amp_date_picker()` 

```php		
static $class_prefixes = [
	'CalendarDay',
	'CalendarMonth',
	'CalendarMonthGrid',
	'DayPicker',
	'DayPickerKeyboardShortcuts',
	'DayPickerNavigation',
	'KeyboardShortcutRow',
];
```
...as they don't begin with `amp-date-picker-`. 

But maybe I'm not understanding your point. And your point about the performance cost is well-taken.

Reply:
Here's an example `<amp-date-picker>` 

<img width="669" alt="amp-date-picker" src="https://user-images.githubusercontent.com/4063887/75833804-c83ea580-5d7f-11ea-80bd-70915b54e143.png">